File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/puzzlegame/Block.java
package com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame;

import android.graphics.Point;
import android.support.annotation.NonNull;

import java.io.Serializable;

/**
 * Created by Chrystian on 2/10/2018.
 */

public class Block implements Serializable, Comparable<Block> {

    private BlockType type;
    private int xCoor;
    private int yCoor;

    public boolean isBeingSwitched;
    public int switchAnimationCount;
    public Direction leftRightAnimationDirection;

    public boolean isAnimatingDown;
    public int downAnimatingCount;

    public boolean hasMatched;
    public int delayMatchAnimationCount;
    public int matchInvertedAnimationCount;
    public int matchPopAnimationCount;
    public int clearMatchCount;
    public int matchTotalCount;

    public boolean hasPopped;
    public int popPosition;

    public boolean canCombo;
    public boolean removeComboFlagOnNextFrame;

    public Block(int t, int x, int y) {
        switch (t) {
            case 1:
                type = BlockType.LEAF;
                break;
            case 2:
                type = BlockType.FIRE;
                break;
            case 3:
                type = BlockType.HEART;
                break;
            case 4:
                type = BlockType.WATER;
                break;
            case 5:
                type = BlockType.COIN;
                break;
            case 6:
                type = BlockType.TRAINER;
                break;
            case 7:
                type = BlockType.DIAMOND;
                break;
            default:
                type = BlockType.EMPTY;
        }

        xCoor = x;
        yCoor = y;

        resetBlockValues();
    }

    public void clear() {
        type = BlockType.EMPTY;
        resetBlockValues();
    }

    private void resetBlockValues() {
        isBeingSwitched = false;
        switchAnimationCount = 0;
        leftRightAnimationDirection = Direction.None;

        isAnimatingDown = false;
        downAnimatingCount = 0;
        hasMatched = false;
        matchPopAnimationCount = 0;
        matchInvertedAnimationCount = 0;
        delayMatchAnimationCount = 0;
        clearMatchCount = 0;
        matchTotalCount = 0;
        hasPopped = false;
        popPosition = 0;

        canCombo = false;
        removeComboFlagOnNextFrame = false;
    }

    @Override
    public int compareTo(@NonNull Block otherBlock) {
        Point p1 = getCoords();
        Point p2 = otherBlock.getCoords();

        if (p1.y == p2.y) {
            return p1.x - p2.x;
        } else {
            return p1.y - p2.y;
        }
    }

    public void changeCoords(int newX, int newY) {
        xCoor = newX;
        yCoor = newY;
    }

    public Point getCoords() {
        return new Point(xCoor, yCoor);
    }

    public boolean isBlockEmpty() {
        return type == BlockType.EMPTY;
    }

    public BlockType getBlockType() {
        return type;
    }

    public void startFallingAnimation() {
        downAnimatingCount = 0;
        isAnimatingDown = true;
    }

    public void stopFallingAnimation() {
        downAnimatingCount = 0;
        isAnimatingDown = false;
    }

    public void startSwitchAnimation(Direction direction) {
        isBeingSwitched = true;
        switchAnimationCount = 1;
        leftRightAnimationDirection = direction;
    }

    public void stopSwitchAnimation() {
        isBeingSwitched = false;
        switchAnimationCount = 0;
        leftRightAnimationDirection = Direction.None;
    }

    public void blockMatched(int delayedMatchAnimationFrames, int matchInvertedAnimationFrames, int clearMatchFrames, int position, int totalNumOfBlockMatches) {
        hasMatched = true;
        matchPopAnimationCount = 0;
        delayMatchAnimationCount = delayedMatchAnimationFrames;
        matchInvertedAnimationCount = matchInvertedAnimationFrames;
        clearMatchCount = clearMatchFrames;
        popPosition = position;
        matchTotalCount = totalNumOfBlockMatches;
    }

    public boolean canInteract() {
        return !isBlockEmpty() && !isAnimating();
    }

    public boolean isAnimating() {
        return isAnimatingDown || hasMatched || isBeingSwitched;
    }

    public void incrementSwitchAnimationFrame() {
        switchAnimationCount++;
    }

    public void incrementDownAnimationFrame() {
        downAnimatingCount++;
    }

    public void incrementPopAnimationFrame() {
        matchPopAnimationCount++;
    }

    public void decrementDelayedMatchAnimationFrame() {
        delayMatchAnimationCount--;
    }

    public void decrementInvertedAnimationFrame() {
        matchInvertedAnimationCount--;
    }

    public void decrementClearFrame() {
        clearMatchCount--;
    }

    public void blockPopped() {
        hasPopped = true;
    }

    public boolean isAnimatingLeft() {
        return leftRightAnimationDirection == Direction.Left;
    }

    public enum BlockType {
        EMPTY(0),
        LEAF(1),
        FIRE(2),
        HEART(3),
        WATER(4),
        COIN(5),
        TRAINER(6),
        DIAMOND(7);

        private final int value;

        BlockType(int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }
    }

}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/puzzlegame/Block.kt
package com.productions.gizzmoo.pokemonpuzzleleague.puzzlegame

import com.productions.gizzmoo.pokemonpuzzleleague.Direction
import java.io.Serializable

class Block(t: Int, x: Int, y: Int) :  Comparable<Block>, Serializable {
    private var xCoord = 0
    private var yCoord = 0

    var type: BlockType
        private set

    var isBeingSwitched: Boolean = false
        private set
    var switchAnimationCount: Int = 0
        private set
    var leftRightAnimationDirection: Direction = Direction.None
        private set

    var isAnimatingDown: Boolean = false
        private set
    var downAnimatingCount: Int = 0
        private set
    var hasMatched: Boolean = false
        private set
    var delayMatchAnimationCount: Int = 0
        private set
    var matchInvertedAnimationCount: Int = 0
        private set
    var matchPopAnimationCount: Int = 0
        private set
    var clearMatchCount: Int = 0
        private set
    var matchTotalCount: Int = 0
        private set
    var hasPopped: Boolean = false
        private set
    var popPosition: Int = 0
        private set

    var canCombo: Boolean = false
        private set
    var removeComboFlagOnNextFrame: Boolean = false
        private set
    var comboCount = 0
        private set

    var maxComboInMatch = 0
        private set

    val canInteract: Boolean
        get() = !isBlockEmpty && !isAnimating

    val isAnimating: Boolean
        get() = isAnimatingDown || hasMatched || isBeingSwitched

    val isBlockEmpty: Boolean
        get() = type === BlockType.EMPTY

    val isAnimatingLeft: Boolean
        get() = leftRightAnimationDirection === Direction.Left

    init {
        type = when (t) {
            1 -> BlockType.LEAF
            2 -> BlockType.FIRE
            3 -> BlockType.HEART
            4 -> BlockType.WATER
            5 -> BlockType.COIN
            6 -> BlockType.DIAMOND
            7 -> BlockType.TRAINER
            else -> BlockType.EMPTY
        }

        xCoord = x
        yCoord = y

        resetBlockValues()
    }

    @Synchronized
    fun clear() {
        type = BlockType.EMPTY
        resetBlockValues()
    }

    override fun compareTo(other: Block): Int {
        return if (this.yCoord == other.yCoord) {
            this.xCoord - other.xCoord
        } else {
            this.yCoord - other.yCoord
        }
    }

    @Synchronized
    fun changeCoords(newX: Int, newY: Int) {
        this.xCoord = newX
        this.yCoord = newY
    }

    @Synchronized
    fun startFallingAnimation() {
        downAnimatingCount = 0
        isAnimatingDown = true
    }

    @Synchronized
    fun stopFallingAnimation() {
        downAnimatingCount = 0
        isAnimatingDown = false
    }

    @Synchronized
    fun startSwitchAnimation(direction: Direction) {
        isBeingSwitched = true
        switchAnimationCount = 1
        leftRightAnimationDirection = direction
    }

    @Synchronized
    fun stopSwitchAnimation() {
        isBeingSwitched = false
        switchAnimationCount = 0
        leftRightAnimationDirection = Direction.None
    }

    @Synchronized
    fun blockMatched(delayedMatchAnimationFrames: Int, matchInvertedAnimationFrames: Int, clearMatchFrames: Int, position: Int, totalNumOfBlockMatches: Int) {
        hasMatched = true
        matchPopAnimationCount = 0
        delayMatchAnimationCount = delayedMatchAnimationFrames
        matchInvertedAnimationCount = matchInvertedAnimationFrames
        clearMatchCount = clearMatchFrames
        popPosition = position
        matchTotalCount = totalNumOfBlockMatches
    }

    @Synchronized
    fun incrementSwitchAnimationFrame() {
        switchAnimationCount++
    }

    @Synchronized
    fun incrementDownAnimationFrame() {
        downAnimatingCount++
    }

    @Synchronized
    fun incrementPopAnimationFrame() {
        matchPopAnimationCount++
    }

    @Synchronized
    fun decrementDelayedMatchAnimationFrame() {
        delayMatchAnimationCount--
    }

    @Synchronized
    fun decrementInvertedAnimationFrame() {
        matchInvertedAnimationCount--
    }

    @Synchronized
    fun decrementClearFrame() {
        clearMatchCount--
    }

    @Synchronized
    fun blockPopped() {
        hasPopped = true
    }

    @Synchronized
    fun setCanComboFlag(flag: Boolean) {
        canCombo = flag
    }

    @Synchronized
    fun setRemoveComboFlagOnNextFrame(flag: Boolean) {
        removeComboFlagOnNextFrame = flag
    }

    @Synchronized
    fun resetComboCount() {
        comboCount = 0
    }

    @Synchronized
    fun setComboCount(count: Int) {
        comboCount = count
    }

    @Synchronized
    fun setMaxComboForMatch(count: Int) {
        maxComboInMatch = count
    }

    private fun resetBlockValues() {
        isBeingSwitched = false
        switchAnimationCount = 0
        leftRightAnimationDirection = Direction.None

        isAnimatingDown = false
        downAnimatingCount = 0
        hasMatched = false
        delayMatchAnimationCount = 0
        matchInvertedAnimationCount = 0
        matchPopAnimationCount = 0
        clearMatchCount = 0
        matchTotalCount = 0
        hasPopped = false
        popPosition = 0

        canCombo = false
        removeComboFlagOnNextFrame = false
        comboCount = 0
        maxComboInMatch = 0
    }
}